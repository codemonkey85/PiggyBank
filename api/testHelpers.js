const testHelpers = {
  normalize: (text) => {
    return text.replace(/\s/g, "").toLowerCase();
  },
  mockRouter: () => {
    return {
      get: jest.fn(),
      post: jest.fn(),
      put: jest.fn(),
      delete: jest.fn()
    };
  },
  mockRepo: () => {
    return {
      account: {
        selectAll: jest.fn(),
        insert: jest.fn(),
        update: jest.fn(),
        delete: jest.fn()
      }
    };
  },
  mockRequest: () => {

  },
  mockResult: () => {
    return {
      json: jest.fn()
    };
  }
};

module.exports = testHelpers;