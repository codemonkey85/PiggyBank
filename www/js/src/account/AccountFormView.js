"use strict";

import { html, render } from "../../lib/lit-html/lit-html.js"
import { AccountCollection } from "./AccountCollection.js"
import { AccountSelectMenuView } from "./AccountSelectMenuView.js"
import { CommoditySelectMenuView } from "../commodity/CommoditySelectMenuView.js";

const template = (d) => html`
<div class="modal fade" id="modalForm" tabindex="-1" role="dialog" aria-hidden="true">
  <div class="modal-dialog modal-xl modal-dialog-centered" role="document">
    <div class="modal-content">
      <div class="modal-body">
        <form class="account-editor form">
          <input id="accountId" type="hidden" value="${d.accountId || ''}" />
          <div class="form-row">
            <div class="form-group col">
              <div id="parentSelect"></div>
            </div>
          </div>
          <div class="form-row">
            <div class="form-group col">
              <div id="commoditySelect"></div>
            </div>
          </div>
          <div class="form-row">
            <div class="form-group col">
              <label for="accountName">Name</label>
              <input id="accountName" class="mr-3 form-control" type="text" value="${d.name || ''}" />
            </div>
          </div>
          <div class="form-row">
            <div class="form-group form-check">
              <div class="col">
                <input id="isPlaceholder" class="form-check-input mr-1" type="checkbox" ?checked=${d.isPlaceholder} />
                <label for="isPlaceholder" class="mr-3">
                  Is Placeholder
                </label>
              </div>
            </div>
          </div>
          <div class="form-row">
            <div class="form-group">
              <button class="save btn btn-success mr-2"><span class="icon-solid-check"></span></button>
              <button class="cancel btn btn-danger"><span class="icon-close"></span></button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

export class AccountFormView extends Backbone.View {
  preinitialize(args) {
    this.tagName = "div"
    this.className = ""

    this.events = {
      "click .btn.save": "save",
      "click .btn.cancel": "cancel"
    };
  }

  save(e) {
    e.preventDefault()

    const accountId = this.$("#accountId").val()
    this.model.set({
      "accountId": accountId ? Number(accountId) : null,
      "currencyId": this.$("#commoditySelect select.select-commodity-id").find(":selected").val(),
      "name": this.$("#accountName").val(),
      "isPlaceholder": this.$("#isPlaceholder")[0].checked ? 1 : 0,
      "parentId": this.$("#parentSelect select.select-account-id").find(":selected").val()
    })

    this.model.save({}, {
      success: () => {
        if (this.model.children === null) {
          this.model.children = new AccountCollection()
        }
        this.close()
        this.trigger("saved")
      },
      error: () => {
        alert("Error saving account")
      }
    })
  }

  cancel(e) {
    e.preventDefault()
    this.close()
  }

  render() {
    render(template(this.model.attributes), this.el)

    const selectMenu = new AccountSelectMenuView()
    selectMenu.render(this.model.get("parentId"), "Parent Account", true)
    this.$("#parentSelect").html(selectMenu.el)

    const commodityMenu = new CommoditySelectMenuView(this.model.get("commodityCollection"))
    commodityMenu.render(this.model.get("currencyId"), "Currency")
    this.$("#commoditySelect").html(commodityMenu.el)

    return this
  }

  close() {
    this.$("#modalForm").modal("hide")
  }
}
