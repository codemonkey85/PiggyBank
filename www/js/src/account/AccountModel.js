"use strict"

export class AccountModel extends Backbone.Model {
  preinitialize() {
    this.urlRoot = "/api/account"
    this.children = []

    this.defaults = {
      name: ""
    }
  }

  longName() {
    let ln = this.get("name")
    let model = this
    while(model.parent) {
      ln = model.parent.get("name") + "::" + ln
      model = model.parent
    }
    return ln
  }
}