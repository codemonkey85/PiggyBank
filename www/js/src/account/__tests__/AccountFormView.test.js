import { mockAccountAssets } from "../../__tests__/testHelpers"
import { AccountFormView } from "../AccountFormView"
import { AccountModel } from "../AccountModel"

let view
beforeEach(() => {
  view = new AccountFormView({ model: window.piggybank.accounts.models[0] })
})

test("AccountFormView has render method", () => {
  const renderedView = view.render()
  expect(view).toBe(renderedView)
  expect(view.el).toMatchSnapshot()
})

test(".close hides the modal and destroys the view", () => {
  const modal = jest.fn()
  view.$ = jest.fn().mockReturnValue({ modal })

  view.close()

  expect(view.$).toHaveBeenCalledWith("#modalForm")
  expect(modal).toHaveBeenCalledWith("hide")
})

test(".cancel prevents default event action and calls .close", () => {
  view.close = jest.fn()
  const preventDefault = jest.fn()

  view.cancel({ preventDefault })

  expect(preventDefault).toHaveBeenCalled()
  expect(view.close).toHaveBeenCalled()
})

test(".save updates the model, closes the form and triggers the 'saved' event", () => {
  const preventDefault = jest.fn()

  view.render()
  view.save({ preventDefault })
})