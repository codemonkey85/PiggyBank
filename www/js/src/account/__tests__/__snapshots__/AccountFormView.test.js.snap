// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountFormView has render method 1`] = `
<div>
  <!---->
  
  
  <div
    aria-hidden="true"
    class="modal fade"
    id="modalForm"
    role="dialog"
    tabindex="-1"
  >
    
    
    <div
      class="modal-dialog modal-xl modal-dialog-centered"
      role="document"
    >
      
      
      <div
        class="modal-content"
      >
        
        
        <div
          class="modal-header"
        >
          
          
          <h3
            class="modal-title"
          >
            Title
          </h3>
          
        
        </div>
        
        
        <form
          class="account-editor form"
        >
          
          
          <div
            class="modal-body"
          >
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group col"
              >
                
                
                <div
                  id="parentSelect"
                >
                  <div>
                    <!---->
                    
  
                    <label
                      for="selectAccountId-test guid"
                    >
                      <!---->
                      Parent Account
                      <!---->
                    </label>
                    
  
                    <select
                      class="select-account-id form-control"
                      id="selectAccountId-test guid"
                    >
                      
    
                      <option
                        value=""
                      >
                        [Choose Account]
                      </option>
                      
  
                      <option
                        value="1"
                      >
                        Assets
                      </option>
                      <option
                        value="3"
                      >
                        Assets::Checking
                      </option>
                      <option
                        value="4"
                      >
                        Assets::Investments
                      </option>
                      <option
                        value="5"
                      >
                        Assets::Investments::401k
                      </option>
                      <option
                        value="2"
                      >
                        Assets::Savings
                      </option>
                      <option
                        value="10"
                      >
                        Liabilities
                      </option>
                      <option
                        value="11"
                      >
                        Liabilities::Car Loan
                      </option>
                      <option
                        value="12"
                      >
                        Liabilities::Mortgage
                      </option>
                    </select>
                    

                    <!---->
                  </div>
                </div>
                
              
              </div>
              
            
            </div>
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group col"
              >
                
                
                <div
                  id="commoditySelect"
                >
                  <div>
                    <!---->
                    
  
                    <label
                      for="selectCommodityId-test guid"
                    >
                      <!---->
                      Currency
                      <!---->
                    </label>
                    
  
                    <select
                      class="select-commodity-id form-control"
                      id="selectCommodityId-test guid"
                    >
                      
    
                      <option
                        value=""
                      >
                        [Choose Commodity]
                      </option>
                      
  
                      <option
                        value="undefined"
                      />
                    </select>
                    

                    <!---->
                  </div>
                </div>
                
              
              </div>
              
            
            </div>
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group col"
              >
                
                
                <label
                  for="accountName"
                >
                  Name
                </label>
                
                
                <input
                  class="mr-3 form-control"
                  id="accountName"
                  type="text"
                  value="Assets"
                />
                
              
              </div>
              
            
            </div>
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group col"
              >
                
                
                <div
                  class="form-check"
                >
                  
                  
                  <input
                    checked=""
                    class="form-check-input"
                    id="isPlaceholder"
                    type="checkbox"
                  />
                  
                  
                  <label
                    class="form-check-label"
                    for="isPlaceholder"
                  >
                    
                    Is Placeholder
                  
                  </label>
                  
                
                </div>
                
              
              </div>
              
            
            </div>
            
          
          </div>
          
          
          <div
            class="modal-footer"
          >
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group"
              >
                
                
                <button
                  class="save btn btn-success mr-2"
                >
                  <span
                    class="icon-solid-check"
                  />
                   Save
                </button>
                
                
                <button
                  class="cancel btn btn-danger"
                >
                  <span
                    class="icon-close"
                  />
                   Cancel
                </button>
                
              
              </div>
              
            
            </div>
            
          
          </div>
          
        
        </form>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  
  

  <!---->
</div>
`;
